text/0000-async-stream.md

```diff
@@ -0,0 +1,907 @@
+
+As defined in the [`Future` docs](https://doc.rust-lang.org/stable/std/future/trait.Future.html):
+
+Once a future has completed (returned Ready from poll), calling its poll method again may panic, block forever, or cause other kinds of problems; the Future trait places no requirements on the effects of such a call. However, as the poll method is not marked unsafe, Rust's usual rules apply: calls must never cause undefined behavior (memory corruption, incorrect use of unsafe functions, or the like), regardless of the future's state.
+
+This RFC takes a similar approach to Streams. In general, once a Stream returns `None`, clients should not poll it again. However, individual streams may choose to resume iteration if the user wishes them to, in which case the stream may or may not eventually yield `Some(Item)` again at some point.
+
+### Usage
+
+A user could create a stream as follows (Example taken from @yoshuawuyt's [implementation pull request](https://github.com/rust-lang/rust/pull/79023)).
+
+Creating a stream involves two steps: creating a `struct` to
+ hold the stream's state, and then implementing [`Stream`] for that
+ `struct`.
+
+ Let's make a stream named `Counter` which counts from `1` to `5`:
+
+```rust
+#![feature(async_stream)]
+# use core::stream::Stream;
+# use core::task::{Context, Poll};
+# use core::pin::Pin;
+
+// First, the struct:
+
+/// A stream which counts from one to five
+struct Counter {
+    count: usize,
+}
+
+// we want our count to start at one, so let's add a new() method to help.
+// This isn't strictly necessary, but is convenient. Note that we start
+// `count` at zero, we'll see why in `poll_next()`'s implementation below.
+impl Counter {
+    fn new() -> Counter {
+        Counter { count: 0 }
+    }
+}
+
+// Then, we implement `Stream` for our `Counter`:
+
+impl Stream for Counter {
+    // we will be counting with usize
+    type Item = usize;
+
+    // poll_next() is the only required method
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Increment our count. This is why we started at zero.
+        self.count += 1;
+
+        // Check to see if we've finished counting or not.
+        if self.count < 6 {
+            Poll::Ready(Some(self.count))
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+```
+
+## next method
+
+We should also implement a next method, similar to [the implementation in the futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+In general, we have purposefully kept the core trait definition minimal. 
+There are a number of useful extension methods that are available, for example, 
+in the `futures-util` crate, but we have not included them because they involve 
+closure arguments, and we have not yet finalized the design of async closures.
+
+However, the core `poll_next` method alone is extremely unergonomic. You can't even iterate 
```

Comment:
This is nit, but this particular line comes across too strongly. I think we can tone it down without losing value:
```suggestion
However, the core `poll_next` method is unergonomic; it does not let you iterate 
```

Reply:
Agreed! Ty!